hw/arm/morello.c

```diff
@@ -0,0 +1,154 @@
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "qemu/osdep.h"
+#include "qemu/units.h"
+#include "qemu/cutils.h"
+#include "qapi/error.h"
+#include "qemu/error-report.h"
+#include "cpu.h"
+#include "hw/boards.h"
+#include "hw/loader.h"
+#include "hw/arm/boot.h"
+#include "cheri_tagmem.h"
+#include "sysemu/sysemu.h"
+#include "hw/arm/armtrickbox.h"
+
+#define RAM_SIZE 0x13000000
+
+#define MORELLO_MAX_CPUS 4
+
+typedef struct MorelloMachineState {
+    /*< private >*/
+    MachineState parent_obj;
+    /*< public >*/
+    struct {
+        ARMCPU core;
+    } cpus[MORELLO_MAX_CPUS];
+} MorelloMachineState;
+
+typedef struct MorelloMachineClass {
+    /*< private >*/
+    MachineClass parent_obj;
+    /*< public >*/
+} MorelloMachineClass;
+
+#define TYPE_MORELLO_MACHINE MACHINE_TYPE_NAME("morello")
+#define MORELLO_MACHINE(obj)                                                   \
+    OBJECT_CHECK(MorelloMachineState, (obj), TYPE_MORELLO_MACHINE)
+
+#define MORELLO_MACHINE_CLASS(klass)                                           \
+    OBJECT_CLASS_CHECK(MorelloMachineClass, (klass), TYPE_MORELLO_MACHINE)
+#define MORELLO_MACHINE_GET_CLASS(obj)                                         \
+    OBJECT_GET_CLASS(MorelloMachineClass, (obj), TYPE_MORELLO_MACHINE)
+
+static void error_and_die(const char *message)
+{
+    error_report("%s\n", message);
+    assert(0);
+    exit(1);
+}
+
+// Lifetime is for the lifetime of the program
+static struct arm_boot_info binfo;
+
+static void morello_machine_init(MachineState *machine)
+{
+    MorelloMachineState *morelloMachineState = MORELLO_MACHINE(machine);
+    Error *errp = NULL;
+
+#ifndef TARGET_CHERI
+#error "Morello board should be used with TARGET_CHERI"
+#endif
+
+    /* Add RAM */
+    memory_region_add_subregion(get_system_memory(), 0, machine->ram);
+
+#ifdef TARGET_CHERI
+    /* Then tag it all */
+    cheri_tag_init(machine->ram, machine->ram_size);
+#endif
+
+    if (strcmp(machine->cpu_type, ARM_CPU_TYPE_NAME("morello")) != 0) {
+        error_and_die(
+            "Use the special 'morello' cpu type for the Morello board.");
+    }
+
+    /* Create CPUS */
+
+    for (size_t i = 0; i != machine->smp.cpus; i++) {
+        object_initialize_child(OBJECT(machine), "cpu[*]",
+                                &morelloMachineState->cpus[i].core,
+                                machine->cpu_type);
+        if (!qdev_realize(DEVICE(&morelloMachineState->cpus[i].core), NULL,
+                          &errp)) {
+            error_report_err(errp);
+            assert(0);
+        }
+    }
+
+    /* Create the trickbox used for testing */
+    arm_trickbox_mm_init_default();
+
+    /* Start logging */
+#ifdef CONFIG_TCG_LOG_INSTR
+    qemu_log_instr_init(CPU(ARM_CPU(first_cpu)));
```

Comment:
This should be called for each CPUState we created above.

Reply:
Done